docs/IR.md

```diff
@@ -1,127 +1,353 @@
+
+### Names Within a Graph
 
 All names MUST adhere to C identifier syntax rules.
 
-#### Nodes
+Names of nodes, inputs, outputs, initializers, and attributes are organized into several namespaces. Within a namespace, each name MUST be unique within that graph. 
+
+The namespaces are:
+
+Namespace|Description
+|---|---|
+Attribute|The names of attributes of an operator. Unique for each operator.
+Value|The names of values – node inputs & outputs, tensor values (if named), graph inputs, outputs.
+Node|The names of graph nodes.
+Graph|The names of graphs within a domain.
+Operator|The names of operators within a domain.
+Shape|The names of tensor shape variables – scoped to the value information records of a graph, which is where shape variables occur.
+
+
+### Nodes
+
+Computation nodes are comprised of a name, the name of an operator that it invokes, a list of named inputs, a list of named outputs, and a list of attributes. 
+
+Input and outputs are positionally associated with operator inputs and outputs. Attributes are associated with operator attributes by name.
+
+They have the following properties:
 
-Each computation node consists of a name, the identifier of the operator to be invoked, a list of named inputs and outputs, and a list of attribute-value pairs.
+Name|Type|Description
+|---|---|---|
+name|string|An optional name of the node, used for diagnostic purposes only.
+input|string[]|Names of the values used by the node to propagate input values to the node operator. It must refer to either a graph input or a node output.
+output|string[]|Names of the outputs used by the node to capture data from the operator invoked by the node. It either introduces a  value in the graph or refers to a graph output.
+op_type|string|The symbolic identifier of the operator to invoke.
+domain|string|The domain of the operator set that contains the operator named by the op_type.
+attribute|Attribute[]|Named attributes, another form of operator parameterization, used for constant values rather than propagated values.
+doc_string|string|A human-readable documentation for this value. Markdown is allowed.
 
 Edges in the computation graph are established by outputs of one node being referenced by name in the inputs of a subsequent node.
 
-Names of tensor values flowing through the graph are unique - a particular name can reference either a graph input or an output of a single node. Reuse of names for several outputs is not allowed.
+Node inputs MAY also refer to graph inputs and initializers. The outputs of a given node MAY introduce new names into the graph, name a graph output, or coincide with the outputs of other nodes. Thus, using overlapping output names, two nodes MAY compute the same output value. Graph outputs MAY NOT be used to establish data dependency edges in the graph by being named as node inputs.
```

Comment:
I don't  understand this ... two different node-outpus cannot have the same name, right?

Reply:
The last sentence "Graph outputs MAY NOT ..." is unclear. Every graph output X must also appear as the output of some node (or must refer to a graph input). This X could be legally used as an input in some other node as well. I guess the sentence is trying to clarify that no dependence edge is created from the graph-output to such an input (in this case)? But it seems to imply the use of X as an input of some node is not valid. Since dependency edges are created only between nodes, perhaps this clarification is not necessary?

Reply:
I think this is clumsy wording on my part. 

The motivation was the If() operator, where you would expect the same outputs, or some overlap of them, to be computed by each of the graph attributes, but it's really the operator outputs that are computed by two graphs, not by two nodes, and graph outputs would be associated with If() outputs positionally, not by name, as far as I understand.

Reply:
Regarding your second comment -- we need to specify that node inputs cannot depend on a graph output that hasn't already been computed by another node. So, I think some clarification is necessary, but you are right in that the current one seems imprecise.